osu.Desktop/Updater/AppImageUpdateManager.cs

```diff
@@ -0,0 +1,551 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.Versioning;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using osu.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
+using osu.Game;
+using osu.Game.Graphics;
+using osu.Game.Online.API;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+using osu.Game.Updater;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Desktop.Updater
+{
+    public static class AppImageUpdateTool
+    {
+        /// <summary>
+        /// Arguments passed to the appimageupdatetool
+        /// <list type="bullet">
+        /// <item><see cref="Check"/><description></description></item>
+        /// <item><see cref="Update"/></item>
+        /// <item><see cref="Overwrite"/></item>
+        /// <item><see cref="ToolVersion"/></item>
+        /// <item><see cref="ToolHelp"/></item>
+        /// </list>
+        /// </summary>
+        public enum ToolArguments
+        {
+            /// <summary>
+            /// Used to check for available updates with the embedded update information within the AppImage
+            /// </summary>
+            Check,
+
+            /// <summary>
+            /// Used to update the AppImage and renaming the old file with a .zs-old suffix
+            /// </summary>
+            Update,
+
+            /// <summary>
+            /// Used to update the AppImage without keeping a backup
+            /// </summary>
+            Overwrite,
+
+            /// <summary>
+            /// Used to get the appimageupdatetool version
+            /// </summary>
+            ToolVersion,
+
+            /// <summary>
+            /// Used to get the appimageupdatetool cli help message
+            /// </summary>
+            ToolHelp
+        }
+
+        /// TODO: Set commandFile to match the shipped binary to prevent searching and remove the unnecessary code
+        private static string commandFile;
+
+        private static readonly string[] toollookupfilenames =
+        {
+            "appimageupdatetool",
+            "appimageupdatetool-x86_64.AppImage"
+        };
+
+        private static string command
+        {
+            get
+            {
+                if (commandFile is default(string))
+                {
+                    foreach (string file in toollookupfilenames)
+                    {
+                        commandFile = file;
+                        if (IsInstalled) break;
+                    }
+                }
```

Comment:
The way this is written is so wrong to me. This needs to be rewritten basically from scratch.

* `commandFile is default(string)` is probably one of the most obfuscated ways to write a null check that I've seen. And `commandFile` should be a `string?` to begin with (the expectation is that all new code should be nullable-annotated). This goes everywhere else, basically every usage of `is default(T)` should be removed.
* `IsInstalled` is shocking. It's a static property, which is derived from the value of `Version`, which is a property that *internally fires a process* via `getOutput()`. This violates all sorts of expectations that reside in idiomatic C# code. Properties should be for simple data access. If you want to fire processes, then please use normal methods to indicate that there may be side effects better. And as mentioned above, please get rid of all of the staticness. There's no real reason for it.